sqlstats: add new MaybeFlushWithDrainer method to persistedSqlStats

kyle-a-wong opened the pull request:
Adds a new SSDrainer interface that PersistedSQLStats uses to flush sql stats.

Instead of calling ConsumeStats on sslocal.SQLStats, persistedsqlstats will use a provided SSDrainer to either DrainStats or Reset.

The ConsumeStats method on sslocal.SQLStats has been removed and its logic has been redistributed to persistedsqlstats, as that was the only consumer of the method. Since all the logic now lives in persistedsqlstats, the flush related functions can be called directly instead of relying on nested callbacks.

Resolves: https://github.com/cockroachdb/cockroach/issues/139143
Epic: CRDB-45771
Release note: None


Note: This is a stack committed, only the last commit in the chain needs to be  reviewed

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140673)
<!-- Reviewable:end -->

kyle-a-wong committed:
sqlstats: add new MaybeFlushWithDrainer method to persistedSqlStats

Adds a new SSDrainer interface that PersistedSQLStats uses to flush
sql stats.

Instead of calling ConsumeStats on sslocal.SQLStats, persistedsqlstats
will use a provided SSDrainer to either DrainStats or Reset.

The ConsumeStats method on sslocal.SQLStats has been removed and its
logic has been redistributed to persistedsqlstats, as that was the
only consumer of the method. Since all the logic now lives in
persistedsqlstats, the flush related functions can be called
directly instead of relying on nested callbacks.

In addition to the new interface, this commit updates the flush
logic to upsert multiple rows in a single upsert statement,
determined by `sql.stats.flush.batch_size`.

Resolves: #139143
Epic: CRDB-45771
Release note: None

dhartunian reviewed (commented):
> Reviewed 41 of 41 files at r1, 3 of 33 files at r2, 10 of 36 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl, @angles-n-daemons, and @xinhaoz)

___
*[`pkg/sql/sqlstats/persistedsqlstats/flush.go` line 114 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJ5FdnX8dRVY5Tl5DDE:-OJ5FdnX8dRVY5Tl5DDF:brm2ca9) ([raw file](https://github.com/cockroachdb/cockroach/blob/eabd876217bac326c6098e95997c8c106f7d4fdb/pkg/sql/sqlstats/persistedsqlstats/flush.go#L114)):*
> ```Go
> 	}
> 
> 	bucket := s.getBucket()
> ```

It's a bit odd that `s` provides the bucket and `ssDrainer` provides the data and we assume that they match. Anything we can do about that? Or maybe there's a connection I'm missing here.
___
*[`pkg/sql/sqlstats/sslocal/sslocal_provider.go` line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJ5GBgoC3SSnDxsezyk:-OJ5GBgoC3SSnDxsezyl:bnozm6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/eabd876217bac326c6098e95997c8c106f7d4fdb/pkg/sql/sqlstats/sslocal/sslocal_provider.go#L149)):*
> ```Go
> 			log.Warningf(ctx, "failed to dump stats to log, %s", err.Error())
> 		}
> 		containerStmtStats, containerTxnStats := container.DrainStats(ctx)
> ```

Is there a way we could say that `DrainStats` gives the caller ownership of these slices? would be nice to to have to copy them over. I assume if we can do that, they're no longer being modified anyway.


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl, @angles-n-daemons, @dhartunian, and @xinhaoz)

___
*[`pkg/sql/sqlstats/persistedsqlstats/flush.go` line 114 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJ5FdnX8dRVY5Tl5DDE:-OJOVG2kE2XLfqDM9i41:b-xkefis) ([raw file](https://github.com/cockroachdb/cockroach/blob/eabd876217bac326c6098e95997c8c106f7d4fdb/pkg/sql/sqlstats/persistedsqlstats/flush.go#L114)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

It's a bit odd that `s` provides the bucket and `ssDrainer` provides the data and we assume that they match. Anything we can do about that? Or maybe there's a connection I'm missing here.
</blockquote></details>

This is how it was previously, I just created a struct to encapsulate this data. That said, the `appstatspb.CollectedStatementStatistics` and `appstatspb.CollectedTransactionStatistics` do have `aggregated_ts` and `aggregation_interval` fields on them already, so theoretically we could populate them when we call `ssmemstorage.Container.DrainStats()`. 

One thing I'm unclear about is the disconnect between collected statistics in the in memory containers and aggregation / bucketing . While the `CollectedStats` related types have these fields, they are only ever set when flushing via persistedstatsSQL, when being queried from `crdb_internal.cluster_statement_statistics`, or requested from one of the gRPC statement APIs. I'm not sure if I'm missing something somewhere, but it seems to me like the determination of which bucket a stat belongs to is tightly coupled to sql stats flush. If flushing is disabled, it seems like queried statistics will always be considered part of the most recent bucket (I don't know if this is correct or not). To me, it would make more sense to push the aggregation data all the way down to the write path of the ssmemstorage.Container, but I haven't spent time evaluation how big of an effort this would be. I don't think it is trivial. 
___
*[`pkg/sql/sqlstats/sslocal/sslocal_provider.go` line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJ5GBgoC3SSnDxsezyk:-OJOqqhW-Edah0nFcayd:b77n6b3) ([raw file](https://github.com/cockroachdb/cockroach/blob/eabd876217bac326c6098e95997c8c106f7d4fdb/pkg/sql/sqlstats/sslocal/sslocal_provider.go#L149)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Is there a way we could say that `DrainStats` gives the caller ownership of these slices? would be nice to to have to copy them over. I assume if we can do that, they're no longer being modified anyway.
</blockquote></details>

As discussed in person:
1. container.DrainStats uses the sqlstats iterators to create new copies of the data and return before deleting the stats from its container. This would be a larger endeavor to re-implement, so I'm gonna punt on that for now.
2. I'll look into the the performance of the operation below where we are flattening the stats using append and spread to see if there are any extra allocations that we may want to avoid.


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl, @angles-n-daemons, @dhartunian, and @xinhaoz)

___
*[`pkg/sql/sqlstats/sslocal/sslocal_provider.go` line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJ5GBgoC3SSnDxsezyk:-OJPgnS25vli5dYiitk5:bsvn7lc) ([raw file](https://github.com/cockroachdb/cockroach/blob/eabd876217bac326c6098e95997c8c106f7d4fdb/pkg/sql/sqlstats/sslocal/sslocal_provider.go#L149)):*
<details><summary><i>Previously, kyle-a-wong (Kyle Wong) wrote…</i></summary><blockquote>

As discussed in person:
1. container.DrainStats uses the sqlstats iterators to create new copies of the data and return before deleting the stats from its container. This would be a larger endeavor to re-implement, so I'm gonna punt on that for now.
2. I'll look into the the performance of the operation below where we are flattening the stats using append and spread to see if there are any extra allocations that we may want to avoid.
</blockquote></details>

I added a new benchmark that benchmarks DrainStats. Locally, I added back in the `ConsumeStats` function updated the benchmark to call `ConsumeStats` ,with noop callbacks, and got the following results:

```
name                             old time/op    new time/op    delta
SqlStatsDrain/drainsql-10-12        469ns ±11%     456ns ±12%     ~     (p=0.382 n=8+9)
SqlStatsDrain/drainsql-50-12       2.12µs ±20%    2.24µs ±13%     ~     (p=0.280 n=10+10)
SqlStatsDrain/drainsql-100-12      4.21µs ±65%    3.44µs ±44%     ~     (p=0.079 n=10+9)
SqlStatsDrain/drainsql-10000-12     356µs ±19%     506µs ±65%     ~     (p=0.113 n=9+10)
SqlStatsDrain/drainsql-1000-12     23.9µs ±11%    28.3µs ±42%  +18.32%  (p=0.011 n=8+9)

name                             old alloc/op   new alloc/op   delta
SqlStatsDrain/drainsql-100-12      17.6kB ± 0%    17.4kB ± 0%   -1.02%  (p=0.000 n=10+10)
SqlStatsDrain/drainsql-10000-12    1.37MB ± 0%    1.36MB ± 0%   -0.96%  (p=0.000 n=10+8)
SqlStatsDrain/drainsql-10-12       1.68kB ± 0%    1.67kB ± 0%   -0.95%  (p=0.000 n=10+10)
SqlStatsDrain/drainsql-50-12       8.74kB ± 0%    8.65kB ± 0%   -0.95%  (p=0.000 n=10+10)
SqlStatsDrain/drainsql-1000-12      180kB ± 0%     178kB ± 0%   -0.91%  (p=0.000 n=10+10)

name                             old allocs/op  new allocs/op  delta
SqlStatsDrain/drainsql-10-12         3.00 ± 0%      3.00 ± 0%     ~     (all equal)
SqlStatsDrain/drainsql-50-12         12.0 ± 0%      12.0 ± 0%     ~     (all equal)
SqlStatsDrain/drainsql-100-12        22.0 ± 0%      22.0 ± 0%     ~     (all equal)
SqlStatsDrain/drainsql-1000-12        203 ± 0%       203 ± 0%     ~     (all equal)
SqlStatsDrain/drainsql-10000-12     1.50k ± 0%     1.50k ± 0%     ~     (all equal)
```

The results show that the new implementation of the DrainStats, that returns the list, results in ~1% more memory consumption than the ConsumeStats. My assumption that this increase in alloc/op is a result of a new slice being created on each append. Do you think this is worth going back to the `ConsumeStats` callback style of processing drained stats?


<!-- Sent from Reviewable.io -->

xinhaoz commented on pkg/sql/sql_activity_update_job_test.go:
```diff
@@ -1046,3 +1044,8 @@ func getStatusJSONProto(
 	return serverutils.GetJSONProto(ts, url, response)
 }
+
+func SqlStatsFlushTestServerLocal(ts serverutils.TestServerInterface) {
```
nit: The function name makes the code a little harder to understand, can we consider renaming or just removing this wrapper?

dhartunian replied:
I'm not sure this function is even necessary. What's the value of hiding what the flush is doing? I'd rather know that `MaybeFlush` is being called in my test without having to hop around.
<!-- Reviewable comment -OJdkU9Y7sUu2-1WFq9I:bma7wn -->
<!-- Sent from Reviewable.io -->

kyle-a-wong replied:
Removed. See other comment for an explanation of why I had originally made these helpers.
<!-- Reviewable comment -OJe5NtGA9G5qdXNvz-o:b-6hp6fw -->
<!-- Sent from Reviewable.io -->

xinhaoz commented on pkg/sql/sqlstats/ssprovider.go:
```diff
@@ -114,3 +114,19 @@ type RecordedTxnStats struct {
 }
+
+// SSDrainer is the interface for draining or resetting sql stats.
+type SSDrainer interface {
```
Can I just double check my understaning for the motivation of these changes (coming from the 1 pager):

What is going to be implementing SSDrainer aside from SQLStats? From my understanding, as part of the new job, we would do the grpc fanout (in some yet to be determined batched method) that will be calling DrainStats on each sslocal.SQLStats container. Where else do we need a drainer?

kyle-a-wong replied:
There will be a new `ssRemote.SQLStats` that will also implement this and will be responsible for the fanout. This is what the job will provide to the `MaybeFlushWithDrainer` method to do the cluster wide drain / persist
<!-- Reviewable comment -OJe0M2fAEmbhzlgpiqt:b-f88o9g -->
<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> Reviewed 1 of 36 files at r3, 2 of 3 files at r4, 4 of 6 files at r5, 2 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl, @angles-n-daemons, @kyle-a-wong, and @xinhaoz)

___
*[`pkg/sql/sqlstats/persistedsqlstats/flush.go` line 382 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJdjnEK0XZNjqnAq4H_:-OJdjnEK0XZNjqnAq4Ha:b-qxbz2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb5dff81dc4222f73c8df7d16b60457b32218fd6/pkg/sql/sqlstats/persistedsqlstats/flush.go#L382)):*
> ```Go
> }
> 
> func generateStatementStatsQuery(statementCount int) string {
> ```

what do you think of just inlining these? unless you're going to write extensive unit tests for them, I'd just inline since args and query need to match up and will be easier to see.
___
*[`pkg/sql/sqlstats/test_utils.go` line 53 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJVGRotEohWRka91eNe:-OJVGRou7ocd3adc9PP2:b-ew7x9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0ad5aabc85d2f9e6089ca6983208dd043f5e4f0/pkg/sql/sqlstats/test_utils.go#L53)):*
> ```Go
> 	SkipZoneConfigBootstrap bool
> 
> 	// FlushInterceptor intercepts persistsqlstats flush operation.
> ```

nit: `persistedsqlstats`
___
*[`pkg/sql/sqlstats/test_utils.go` line 20 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJdiFOB9HsC39yPADSg:-OJdiFOB9HsC39yPADSh:b-mqput6) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb5dff81dc4222f73c8df7d16b60457b32218fd6/pkg/sql/sqlstats/test_utils.go#L20)):*
> ```Go
> 	aggregatedTs time.Time,
> 	stmtStats []*appstatspb.CollectedStatementStatistics,
> 	txnStats []*appstatspb.CollectedTransactionStatistics)
> ```

nit: comma + line break before `)`. our formatter doesn't do that?
___
*[`pkg/sql/sqlstats/persistedsqlstats/provider.go` line 34 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJdkLcEECp5HuOYlUQr:-OJdkLcEECp5HuOYlUQs:b-ea6kcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb5dff81dc4222f73c8df7d16b60457b32218fd6/pkg/sql/sqlstats/persistedsqlstats/provider.go#L34)):*
> ```Go
> )
> 
> const statementStatisticUpsertQuery = `
> ```

why do these live so far away from where they're used? can we move next to usage?
___
*[`pkg/sql/sqlstats/persistedsqlstats/sqlstatstestutil/testutils.go` line 298 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJVHpgq3slBXlJNA5KK:-OJVHpgq3slBXlJNA5KL:b-tt8e93) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0ad5aabc85d2f9e6089ca6983208dd043f5e4f0/pkg/sql/sqlstats/persistedsqlstats/sqlstatstestutil/testutils.go#L298)):*
> ```Go
> }
> 
> func FlushTestServerLocal(ts serverutils.TestServerInterface) {
> ```

@kyle-a-wong what does this helper provide? I think for the long line that it saves it also obscures something quite simple. I'd just leave the original call here unless there's a compelling reason I'm missing.


<!-- Sent from Reviewable.io -->

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140673)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl, @angles-n-daemons, @dhartunian, and @xinhaoz)

___
*[`pkg/sql/sqlstats/test_utils.go` line 53 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJVGRotEohWRka91eNe:-OJe1v6D9Jgq6WhLbAIL:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0ad5aabc85d2f9e6089ca6983208dd043f5e4f0/pkg/sql/sqlstats/test_utils.go#L53)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: `persistedsqlstats`
</blockquote></details>

done
___
*[`pkg/sql/sqlstats/test_utils.go` line 20 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJdiFOB9HsC39yPADSg:-OJe1e0zAj6G7rhKzsgQ:bs9l50i) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb5dff81dc4222f73c8df7d16b60457b32218fd6/pkg/sql/sqlstats/test_utils.go#L20)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: comma + line break before `)`. our formatter doesn't do that?
</blockquote></details>

I guess not. Done
___
*[`pkg/sql/sqlstats/persistedsqlstats/flush.go` line 382 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJdjnEK0XZNjqnAq4H_:-OJe2wOoDs2ae-ptf1ey:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb5dff81dc4222f73c8df7d16b60457b32218fd6/pkg/sql/sqlstats/persistedsqlstats/flush.go#L382)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

what do you think of just inlining these? unless you're going to write extensive unit tests for them, I'd just inline since args and query need to match up and will be easier to see.
</blockquote></details>

Done
___
*[`pkg/sql/sqlstats/persistedsqlstats/provider.go` line 34 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJdkLcEECp5HuOYlUQr:-OJe3-gg0dUFQS0p0eCh:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb5dff81dc4222f73c8df7d16b60457b32218fd6/pkg/sql/sqlstats/persistedsqlstats/provider.go#L34)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

why do these live so far away from where they're used? can we move next to usage?
</blockquote></details>

Done
___
*[`pkg/sql/sqlstats/persistedsqlstats/sqlstatstestutil/testutils.go` line 298 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-OJVHpgq3slBXlJNA5KK:-OJe3yMZBkGLRoxc2u9P:b-r8qdnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0ad5aabc85d2f9e6089ca6983208dd043f5e4f0/pkg/sql/sqlstats/persistedsqlstats/sqlstatstestutil/testutils.go#L298)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

@kyle-a-wong what does this helper provide? I think for the long line that it saves it also obscures something quite simple. I'd just leave the original call here unless there's a compelling reason I'm missing.
</blockquote></details>

Originally, this PR contained a larger refactor that included getting rid of the embded `sslocal.SQLStats` from `persistedSQLStats` and changed the `MaybeFlush` signature to require the SSDrainer to be  provided. Doing that required all updates to all tests t hat called `MaybeFlush` to get this `localSqlStats` object from the server and it felt cleaner to me to only have to do that in one place instead of every place we call `MaybeFlush`.

I've since went back on that idea for this PR and opted for the new `MaybeFlushWithDrainer` method instead, but i never removed these helps / the refactor. If necessary I can add this back in later, but for now i've removed this helper


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
Nice work. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Can you link to https://github.com/cockroachdb/cockroach/issues/139143 which I assigned to you prior to merging?

> Reviewed 6 of 13 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140673#-:-OJeTvy54VCZPrg13u-g:b-dbeutg)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @alyshanjahani-crl, @angles-n-daemons, and @xinhaoz)



<!-- Sent from Reviewable.io -->

kyle-a-wong commented:
bors retry
